### ISSUE TITLE
`TableHelperMacro` sometimes unable to discover table ancestor type

### ISSUE THREAD

**Issue by OlegIlyenko** [outworkers/phantom#577]
Looks like a race condition in scalac of some sort. It compiles perfectly in my dev environment, but **sometimes** fails to compile on our CI server.

I get following error message from scala compiler:

```bash
CassandraCommitModel.scala:11: Could not find a direct ancestor for type CassandraCommitModel
 abstract class CassandraCommitModel(eventSerializers: Map[String, JSON[_]]) extends CassandraTable[CassandraCommitModel, Commit] with RootConnector  {
```

I would appreciate any help with this issue. Maybe I'm doing something wrong? (looks like macro was introduced pretty recently, should I change the table definition in some way?)

**Comment by alexflav23** [comment]
Hi @OlegIlyenko,

Sounds like this is indeed a problem, the macro was introduced very recently yes, 2.0.x is not officially launched as far as we are concerned, you are taking releases of a branch.

Now of course we welcome bug reports, could you please show me the full class structure of how you are constructing your tables? I don't care about anything speciic to your company, more about the class structure, since the macro can theoretically choke up in certain complex inheritance patterns, it has to find a particular ancestor for things to work.

Regards,
Flavian

**Comment by OlegIlyenko** [comment]
Thanks for the very fast rely!

I forgot to mention the versions:

* `phantom-dsl`: 1.29.5
* `scala`: 2.11.8
* `java`:  1.8.0_72-b15

Here is the fill model definition:

```scala
package foo

import java.util.UUID
import com.websudos.phantom.dsl._
import org.joda.time.DateTimeZone
// more imports...

abstract class CassandraCommitModel(eventSerializers: Map[String, JSON[_]]) extends CassandraTable[CassandraCommitModel, Commit] with RootConnector  {
  override def tableName: String = "commits"

  def json[T : JSON] = new JsonColumn[T](this) {
    override def fromJson(obj: String): T =
      getFromJSON[T](obj)

    override def toJson(obj: T): String =
      toJSON(obj)
  }

  object id extends UUIDColumn(this) with PartitionKey[UUID]
  // more columns...

  override def fromRow(r: Row): Commit = {
    // ...
  }
}
```

and the database definition:

```scala
package bar

import foo.CassandraCommitModel
import com.datastax.driver.core.{Session, SimpleStatement}
import com.websudos.phantom.builder.query.prepared.ExecutablePreparedQuery
import com.websudos.phantom.builder.query.{CQLQuery, ExecutableStatement, ExecutableStatementList, QueryOptions}
import com.websudos.phantom.connectors.KeySpace
import com.websudos.phantom.database.Database
import com.websudos.phantom.dsl._
// more imports...

class CassandraDatabase(override val connector: KeySpaceDef, eventSerializers: Map[String, JSON[_]]) extends Database[CassandraDatabase](connector) {
  object commits extends CassandraCommitModel(eventSerializers) with connector.Connector

  def createKeySpace(keySpace: KeySpace, replication: String) = // ...

  def deleteKeySpace(keySpace: KeySpace) = // ...

  def truncateKeySpace(keySpace: KeySpace) = // ...
    
}

object CassandraDatabase {
  // more things here
}
```

let me know if you need more info

**Comment by alexflav23** [comment]
1.29.5 doesn't use macros, you are on 2.0.+? Please double check that.

**Comment by alexflav23** [comment]
Hi @OlegIlyenko,

Ok my bad, an incomplete partial release made its way live through a CI error we forgot to address. I'd go back to 1.29.4 or up to 2.0.+, but that 2.0 upgrade is a bit more effort. There will be an upgrade guide out shortly.

Regards,
Flavian

**Comment by OlegIlyenko** [comment]
here is an evidence :)

![tablehelper_scala_-_intellij_idea_2016_2_2](https://cloud.githubusercontent.com/assets/156569/20141765/1b58deaa-a692-11e6-9df8-ababa84c0e6e.png)

**Comment by alexflav23** [comment]
HI @OlegIlyenko,

Yep, greedy CI published a release that was never intended to go out, I forgot about that for a second :)

Regards.

**Comment by OlegIlyenko** [comment]
Great, thanks a lot for looking into it! Looking forward to the next release :) I will go back to `1.29.4` for the moment being.

**Comment by alexflav23** [comment]
Hi @OlegIlyenko, I have also found this error when I mark the table class as `abstract`. I would also expect it to fail when you use inheritance patterns and tables extend each other. It's meant to find the first direct ancestor of `CassandraTable` that doesn't also mix in `RootConnector`.

The reason for this is basically because we need to perform an `.asInstanceOf` cast, F bounded type polymorphism in Scala is not directly encoded. `trait Test[T <: Test] { val t = this.asInstanceOf[T]}`, otherwise references are not found.

Maybe this will be fixed in Scala itself, that would certainly make life a lot easier.

**Comment by OlegIlyenko** [comment]
hmm... I wonder how it will look in scala 2.12. During the 2.12 migration I noticed that macro behavior changed in some places. Some things also became a bit more stable.

**Comment by alexflav23** [comment]
Hi @OlegIlyenko,

That's correct, it may be that 2.12 will fix a whole lot of issues though I'd imagine this won't be a deal breaker. Now, a simple way to fix your problem is to get rid of the inheritance, so instead of having `Table` and `ConcreteTable` just have the one, that seems to fix it relatively quickly.

At the very least as an interim solution, that should hopefully help you fix things. There are also published versions of the 2.0.x series on our `Resolver.bintrayRepo("outworkers", "oss-releases")`, but that's more migration work as a lot has changed.

The intial [2.0.0 migration guide](https://github.com/outworkers/phantom/tree/feature/2.0.0#200-migration-guide) can help you get started.

Regards,
Flavian

**Comment by lahirug** [comment]
I am using 2.0.9 and I still get this error,  My table definition looks like this.

```
case class ServerSessionTimeSeriesKey(sessionId: UUID, serverId: String, serverVersion: String, key: String)

abstract class ServerSessionKeysBySessionUuid(implicit ks: KeySpace)
    extends CassandraTable[ServerSessionKeysBySessionUuid, ServerSessionTimeSeriesKey] with RootConnector {


```

**Comment by alexflav23** [comment]
It's the `abstract` that causes this, though we don't know why yet. In the interim, could you remove it somehow?

**Comment by lahirug** [comment]
No its not fixing it and I see that its intermittently giving me this error.

**Comment by alexflav23** [comment]
Hi @lahirug, is the above the full code example I need to replicate the error? I'll have a look

**Comment by alexflav23** [comment]
HI @OlegIlyenko @lahirug this has been resolved in `2.0.0`. A `1.30.x` release will be out later today or tomorrow to remove the unstable macro component that accidentally leaked thanks to CI, but that aside this has been properly addressed in `2.0.11`.

I am going to close this ticket as the macros were never intended to make their way here, and we are not planning to backport the macro part onto Phantom `1.x.x`.
